arch/arm/core/userspace.S

```diff
@@ -0,0 +1,154 @@
+/*
+ * Userspace and service handler hooks
+ *
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+#include <offsets_short.h>
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <kernel_structs.h>
+#include <arch/cpu.h>
+#include <syscall.h>
+
+_ASM_FILE_PROLOGUE
+
+GTEXT(_arm_userspace_enter)
+GTEXT(_arm_do_syscall)
+GDATA(_kernel)
+
+/* Imports */
+GTEXT(_k_syscall_table)
+
+/**
+ *
+ * User space entry function
+ *
+ * This function is the entry point to user mode from privileged execution.
+ * The conversion is one way, and threads which transition to user mode do
+ * not transition back later, unless they are doing system calls.
+ *
+ */
+SECTION_FUNC(TEXT,_arm_userspace_enter)
+    /* preserve arguments */
+    mov lr,r0
+    push {r1,r2,r3,lr}
+
+    /* disable mpu stack guard to keep us from triggering guard region */
+    bl arm_core_mpu_disable
+    pop {r1,r2,r3,lr}
+
+    /*
+     * calculate end of stack and set SP to it
+     */
+    ldr r0, [sp]
+    ldr ip, [sp,#4]
+    add r0,ip
+
+    msr PSP, r0
+
+    /* push args to stack, pad end with 2 words of 0x0 */
+    sub sp,8
+    push {r1,r2,r3,lr}
+    ldr r3,=0
+    str r3,[sp, #20]
+    str r3,[sp, #24]
+
+    /* clear the stack area to clean out privileged data */
+#ifdef CONFIG_INIT_STACKS
+    ldr r1,=0xaaaaaaaa
+#else
+    ldr r1,=0
+#endif
+    mov r2,ip
+    sub r2,#24		/* don't erase the pushed regs */
+    bl memset
+
+#ifdef CONFIG_MPU_STACK_GUARD
+    /* setup arguments to thread_entry function */
+    ldr r0,=_kernel
+    ldr r0,[r0, #_kernel_offset_to_current]
+    bl configure_mpu_stack_guard
```

Comment:
I think the stack guard has been configured in __pendsv before _arch_user_mode_enter() being called.

Reply:
can we just make any adjustments needed to the guard area in C code and not in here?

i also don't have a clear picture on how the guard area, kernel stack, and thread stack are arranged

Reply:
This gets complicated, because we have to use the guard region differently based on if we are in userspace or not (regardless of if STACK_GUARD is defined).  Basically from lowest to highest address the memory is laid out like:
|-----guard------|---------stack-----------------|
with userspace it'll be:
|------guard-----|-------------privileged stack------|-------unprivileged stack--------|

Added complexity is that the mpu regions have to be aligned.  So this means when we do userspace we absolutely have to have the stacks aligned to power of 2 boundaries.  This means the guard HAS to come out of the stack, not added to the stack.  All stacks have to fit in a power of 2 size.  The privileged stack also has to be power of 2.

About to push some fixes for this that changes the mpu regions to work properly for user space